File: SimpleAccounting/app/src/main/java/com/emmanuelmess/simpleaccounting/dataloading/LoadMonthAsyncTask.java
package com.emmanuelmess.simpleaccounting.dataloading;

import android.os.AsyncTask;
import android.util.Pair;

import com.emmanuelmess.simpleaccounting.db.TableGeneral;

import java.util.ArrayList;

/**
 * @author Emmanuel
 *         on 27/11/2016, at 15:03.
 */

public class LoadMonthAsyncTask extends AsyncTask<Void, Void, Pair<String[][], ArrayList<Integer>>> {

	private int month, year;
	private String currency;
	private TableGeneral tableGeneral;
	private AsyncFinishedListener<Pair<String[][], ArrayList<Integer>>> listener;
	private static boolean alreadyLoading = false;

	public LoadMonthAsyncTask(int m, int y, String c, TableGeneral dbG,
							  AsyncFinishedListener<Pair<String[][], ArrayList<Integer>>> l) {
		month = m;
		year = y;
		currency = c;
		tableGeneral = dbG;
		listener = l;
	}

	@Override
	protected Pair<String[][], ArrayList<Integer>> doInBackground(Void... p) {
		if(!alreadyLoading)
			alreadyLoading = true;
		else throw new IllegalStateException("Already loading month: " + year + "-" + (month+1));

		int[] data = tableGeneral.getIndexesForMonth(month, year, currency);
		ArrayList<Integer> rowToDBRowConversion = new ArrayList<>();

		if(isCancelled()) return null;

		for(int m : data)
			rowToDBRowConversion.add(m);

		if(isCancelled()) return null;

		return new Pair<>(tableGeneral.getAllForMonth(month, year, currency), rowToDBRowConversion);
	}

	@Override
	protected void onPostExecute(Pair<String[][], ArrayList<Integer>> dbRowsPairedRowToDBConversion) {
		if(!isCancelled())
			listener.OnAsyncFinished(dbRowsPairedRowToDBConversion);

		alreadyLoading = false;
	}

	public static boolean isAlreadyLoading() {
		return alreadyLoading;
	}

}

File: SimpleAccounting/app/src/main/java/com/emmanuelmess/simpleaccounting/dataloading/async/LoadMonthAsyncTask.kt
package com.emmanuelmess.simpleaccounting.dataloading.async

import android.os.AsyncTask
import com.emmanuelmess.simpleaccounting.dataloading.data.MonthData
import com.emmanuelmess.simpleaccounting.dataloading.data.Session

import com.emmanuelmess.simpleaccounting.db.TableGeneral
import com.emmanuelmess.simpleaccounting.db.TableMonthlyBalance

import java.util.ArrayList

/**
 * @author Emmanuel
 * on 27/11/2016, at 15:03.
 */

class LoadMonthAsyncTask(
	private val session: Session,
	private val tableMonthlyBalance: TableMonthlyBalance?,
	private val tableGeneral: TableGeneral,
	private val listener: AsyncFinishedListener<MonthData>
) : AsyncTask<Void, Void, MonthData>() {

	@Suppress("UsePropertyAccessSyntax")
	override fun onPreExecute() {
		tableMonthlyBalance?.getReadableDatabase()//updates the database, calls onUpgrade()
		tableGeneral.getReadableDatabase()//triggers onUpdate()

	}

	override fun doInBackground(vararg p: Void): MonthData? {
		if (!isAlreadyLoading)
			isAlreadyLoading = true
		else
			throw IllegalStateException("Already loading month: ${session.year}-${session.month + 1}");

		val data = tableGeneral.getIndexesForMonth(session)
		val rowToDBRowConversion = ArrayList<Int>()

		if (isCancelled) return null

		for (m in data)
			rowToDBRowConversion.add(m)

		return (
			if (isCancelled) null
			else MonthData(
				session,
				tableMonthlyBalance?.getBalanceLastMonthWithData(session),
				tableGeneral.getAllForMonth(session),
				rowToDBRowConversion)
		)
	}

	override fun onPostExecute(dbRowsPairedRowToDBConversion: MonthData) {
		if (!isCancelled)
			listener.onAsyncFinished(dbRowsPairedRowToDBConversion)

		isAlreadyLoading = false
	}

	override fun onCancelled(result: MonthData) {
		isAlreadyLoading = false
	}

	companion object {
		var isAlreadyLoading = false
			private set
	}

}
